File: src/lab02/libs/matrixes.js
import * as arrays from "./arrays.js"

export function transpose(a){
    //exceptions
    if (!(a instanceof Array)) throw new TypeError("TypeError");
    if (a.length === 0) return [];
    if (a.length !== 0) var l = a[0].length
    for (var i of a) if (i.length != l) throw new TypeError("ValueError");

    //actual code
    var ans = [];
    var m = a.length;
    var n = a[0].length;
    for (var i = 0; i < n; i++)
    {
        let temp = []
        for (var j = 0; j < m; j++)
            temp.push(a[j][i]);
        ans.push(temp);
        temp = [];
    }
    return ans;
}


export function row_sums(a){
    //exceptions
    if (!(a instanceof Array)) throw new TypeError("TypeError");
    if (a.length === 0) return [];
    if (a.length !== 0) var l = a[0].length
    for (var i of a) if (i.length != l) throw new TypeError("ValueError");

    //actual code
    var ans = []
    for (var i of a) {
        var sum = 0;
        for (var j of i)
            sum+=j;
        ans.push(sum)
    }
    return ans

}

export function col_sums(a){
    //exceptions
    if (!(a instanceof Array)) throw new TypeError("TypeError");
    if (a.length === 0) return [];
    if (a.length !== 0) var l = a[0].length
    for (var i of a) if (i.length != l) throw new TypeError("ValueError");

    //actual code
    var ans = []
    for (var m = 0; m < a[0].length; m++)
    {
        var temp = 0;
        for (var n = 0; n < a.length; n++)
            temp += a[n][m];
        ans.push(temp);
    }
    return ans

}

File: src/lib/matrixes.py
def transpose(a):
    # exceptions
    if not isinstance(a, list):
        raise TypeError("TypeError")
    if len(a) == 0:
        return []
    if len(a) != 0:
        l = len(a[0])
    for i in a:
        if len(i) != l:
            raise TypeError("ValueError")

    # actual code
    ans = []
    m = len(a)
    n = len(a[0])
    for i in range(n):
        temp = []
        for j in range(m):
            temp.append(a[j][i])
        ans.append(temp)
        temp = []
    return ans


def row_sums(a):
    # exceptions
    if not isinstance(a, list):
        raise TypeError("TypeError")
    if len(a) == 0:
        return []
    l = len(a[0])
    for i in a:
        if len(i) != l:
            raise TypeError("ValueError")

    # actual code
    ans = []
    for i in a:
        sum = 0
        for j in i:
            sum += j
        ans.append(sum)
    return ans


def col_sums(a):
    # exceptions
    if not isinstance(a, list):
        raise TypeError("TypeError")
    if len(a) == 0:
        return []
    l = len(a[0])
    for i in a:
        if len(i) != l:
            raise TypeError("ValueError")

    # actual code
    ans = []
    for m in range(len(a[0])):
        temp = 0
        for n in range(len(a)):
            temp += a[n][m]
        ans.append(temp)
    return ans
